feature/anton_apr1

anton-anufriev opened the pull request:
Added Log, LogEntry, CommandLogEntry
Change CommandMessage. Added AppendRequestHandler for all states and overrode logic for Candidate state.
De-inherited DirectCommand from Message.
Added replayEvent flag to Transition.

anton-anufriev committed:
1. append request in candidate state, however should be reused for all
2. Persistent state to get termId at index
3. Persistent statte and messageLog to truncate
... work in progress

anton-anufriev committed:
Added Log, LogEntry, CommandLogEntry
Change CommandMessage. Added AppendRequestHandler for all states and overrode logic for Candidate state.
 De-inherited DirectCommand from Message.
 Added replayEvent flag to Transition.

terzerm commented on src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java:
```diff
@@ -0,0 +1,70 @@
+
+public class AppendRequestHandler {
+    private final PersistentState persistentState;
+    private final VolatileState volatileState;
```
~~you don't need those member variables cause they are passed to the onAppendRequest method via ServerContext~~

terzerm commented on src/main/java/org/tools4j/hoverraft/event/AppendRequestHandler.java:
```diff
@@ -0,0 +1,70 @@
+        final CommandLogEntry newCommandLogEntry = appendRequest.commandLogEntry();
+        final CommandLog commandLog = persistentState.commandLog();
+
+        CommandLog.CONTAINMENT containment =  commandLog.contains(prevLogEntry);
```
final

terzerm commented on src/main/java/org/tools4j/hoverraft/message/MessageType.java:
```diff
@@ -35,7 +36,7 @@ public Message create(final MessageFactory factory) {
         }
         @Override
-        public DirectMessage create(final DirectMessageFactory factory) {
+        public AbstractDirectMessage create(final DirectMessageFactory factory) {
```
I don't like if abstract types are exposed anywhere

anton-anufriev replied:
Yeah, I did think about that and left it for discussion. Fixed now.

terzerm commented on src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java:
```diff
@@ -93,39 +107,41 @@ public AppendRequest leaderId(final int leaderId) {
-        return this;
+    public void wrap(DirectBuffer buffer, int offset) {
+        super.wrap(buffer, offset);
+        directCommandLogEntry.wrap(buffer, offset + COMMAND_LOG_ENTRY_OFF);
```
nice!

terzerm commented on src/main/java/org/tools4j/hoverraft/message/direct/DirectAppendRequest.java:
```diff
@@ -93,39 +107,41 @@ public AppendRequest leaderId(final int leaderId) {
+    @Override
+    public void unwrap() {
+        super.unwrap();
+        directCommandLogEntry.unwrap();
```
should be opposite unwrapping order imo

anton-anufriev replied:
fixed

terzerm commented on src/main/java/org/tools4j/hoverraft/message/direct/DirectMessageFactory.java:
```diff
@@ -84,11 +84,11 @@ public static DirectMessageFactory createForWriting(final MutableDirectBuffer bu
         if (type >= 0 & type <= MessageType.maxOrdinal()) {
             final MessageType messageType = MessageType.valueByOrdinal(type);
-            final DirectMessage message = messageType.create(this);
+            final AbstractDirectMessage message = messageType.create(this);
```
don't like that abstractXXXmessage is exposed

anton-anufriev replied:
fixed

terzerm commented on src/main/java/org/tools4j/hoverraft/state/CandidateState.java:
```diff
@@ -81,26 +81,17 @@ private Transition onVoteResponse(final ServerContext serverContext, final VoteR
         } else {
-            successful = false;
-            transition = Transition.STEADY;
+            return super.onAppendRequest(serverContext, appendRequest);
```
not sure I like that

anton-anufriev replied:
why?

terzerm commented on src/main/java/org/tools4j/hoverraft/state/CommandLog.java:
```diff
@@ -0,0 +1,12 @@
+package org.tools4j.hoverraft.state;
+
+public interface CommandLog extends LogEntry, Comparable<LogEntry> {
```
it is dodgy that CommandLog _is a_ LogEntry

anton-anufriev replied:
Agree, fixed.

terzerm commented on src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java:
```diff
@@ -0,0 +1,106 @@
+        if (readIndex.get() < commandLogEntries.size()) {
+            return commandLogEntries.get(readIndex.getAndIncrement());
+        }
+        throw new IllegalArgumentException("Read index " + readIndex + " has reached end of message log with size " + commandLogEntries.size());
```
there are no arguments. this is either IllegalStateEx or it should never be possible to get into this situation.

It should actually only occur if index=0 and log is empty. I would throw an IllegalStateException in this case

anton-anufriev replied:
Copied from your InMemoryMessageLog. Agree that IllegalStateEx is better.

anton-anufriev replied:
Fixed in both

terzerm commented on src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    public synchronized void append(final CommandLogEntry commandLogEntry) {
+        commandLogEntries.add(Objects.requireNonNull(commandLogEntry));
```
maybe we should clone here somehow

anton-anufriev replied:
I think it is better that InMemoryFactory with InMemoryMessages

terzerm commented on src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java:
```diff
@@ -0,0 +1,106 @@
+
+
+    @Override
+    public synchronized void truncate(long index) {
```
method name could be clearer, e.g. `truncateIncluding(..)`

anton-anufriev replied:
done

terzerm commented on src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java:
```diff
@@ -0,0 +1,106 @@
+            } else {
+                return CONTAINMENT.IN;
+            }
+
```
minor: extra empty lines

anton-anufriev replied:
easy

terzerm commented on src/main/java/org/tools4j/hoverraft/state/InMemoryCommandLog.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public int compareTo(LogEntry logEntry) {
```
I think CommandLog should have a method
`LogEntry lastEntry()`

and not implement LogEntry and also not be comparable to LogEntry.

LogEntry itself then implements Comparable<LogEntry>

Note that `index()` should then probably become `lastIndex()` or be removed

anton-anufriev replied:
Yes, I have done it.

terzerm commented on src/main/java/org/tools4j/hoverraft/state/Transition.java:
```diff
@@ -31,21 +31,27 @@
         return this == STEADY ? currentRole : targetRole;
     }
 
+    public final boolean replayEvent() {
```
nice. I like this change.

terzerm commented on src/main/java/org/tools4j/hoverraft/state/direct/DirectCommandLogEntry.java:
```diff
@@ -0,0 +1,93 @@
+
+import java.util.Objects;
+
+public class DirectCommandLogEntry implements CommandLogEntry, DirectMessage {
```
nice. I like this class.

terzerm commented on src/main/java/org/tools4j/hoverraft/state/direct/DirectPersistentState.java:
```diff
@@ -62,8 +63,10 @@ public int votedFor() {
+    @Override
+    public CommandLog commandLog() {
+        //FIXme
+        return null;
```
throwing an exception is better

anton-anufriev replied:
nice! done

terzerm commented on src/main/java/org/tools4j/hoverraft/transport/aeron/AeronReceiver.java:
```diff
@@ -27,14 +27,15 @@
 import java.util.function.Consumer;
 
-public class AeronReceiver implements Receiver<DirectMessage> {
+public class AeronReceiver implements Receiver<AbstractDirectMessage> {
```
no way --- i really don't like that AbstractDirectMessage is now exposed everywhere.

anton-anufriev replied:
it was a temporary change. Fixed

terzerm commented on src/main/java/org/tools4j/hoverraft/transport/chronicle/ChronicleMessageLog.java:
```diff
@@ -140,4 +140,10 @@ private Method initLastWrittenIndexMethod() {
+    @Override
+    public long truncate(long inclusiveIndex) {
+        //FIXme implement it
+        return 0;
```
throw ex

terzerm commented on src/main/java/org/tools4j/hoverraft/transport/embedded/InMemoryMessageLog.java:
```diff
@@ -81,4 +81,10 @@ public synchronized M read() {
+    @Override
+    public long truncate(long inclusiveIndex) {
+        //FIXme implement it
+        return 0;
```
throw ex

terzerm commented on src/test/java/org/tools4j/hoverraft/server/FollowerStateTest.java:
```diff
@@ -68,16 +68,33 @@ public void onAppendRequest() throws Exception {
+        appendRequest.prevLogEntry().term(term)
+                                    .index(100);
+
+        when(persistentState.commandLog().contains(appendRequest.prevLogEntry())).thenReturn(CommandLog.CONTAINMENT.IN);
```
extract commandLog and prefLogEntry into variable for readability

anton-anufriev replied:
done

terzerm commented on src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java:
```diff
@@ -177,13 +186,15 @@ public void onVoteRequest_wrongTerm() throws Exception {
+                .candidateId(serverId);
+        //need LogEntry mutators to return parent object on done()
+        voteRequest.lastLogEntry().term(lastLogTerm)
+                .index(lastLogIndex);
```
that's the disadvantage of the nested object structure --- the method chaining breaks down :-(

terzerm commented on src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java:
```diff
@@ -223,13 +235,15 @@ public void onVoteRequest_invalidCandidate_badLastLogIndex() throws Exception {
         when(persistentState.votedFor()).thenReturn(PersistentState.NOT_VOTED_YET);
-        when(persistentState.lastLogTerm()).thenReturn(lastLogTerm);
-        when(persistentState.lastLogIndex()).thenReturn(lastLogIndex);
+        when(persistentState.commandLog().compareTo(voteRequest.lastLogEntry())).thenReturn(1);
```
extract commandLog and lastLogEntry into variable for readability

anton-anufriev replied:
done

terzerm commented on src/test/java/org/tools4j/hoverraft/state/InMemoryCommandLogTest.java:
```diff
@@ -0,0 +1,29 @@
+        assertThat(commandLog.size()).isEqualTo(1);
+
+
+    }
```
poor test coverage

terzerm reviewed (changes requested):
Must fixes in my opinion:
- exposure of AbstractDirectMessage
- CommandMessageLog implementing LogEntry --- it should be _has a_ instead of _is a_ relation

anton-anufriev committed:
Removed PersistentState and VolatileState from all State implementations and added their accessors to the ServerContext, so that any handler can always access the PersistentState and VolatileState via ServerContext.
Hence, some bits of code got simplified a bit.
Fixed leaky interface ServerContext to return MessageFactory rather than DirectMessageFactory.

anton-anufriev committed:
post review changes, new DirectLogEntry and reusing it in appendRequest.prevLogEntry, voteRequest.lastLogEntry and CommandLogEntry

anton-anufriev committed:
removed AbstractMessage and added AbstractDirectPayload
fixed offset bugs in wrap() methods
moved contains() body to the interface as default method as it is an implementation agnostic body

anton-anufriev committed:
implemented InMemoryMessageFactory
added some direct messages tests

anton-anufriev committed:
Reverted to have persistentState and volatileState in States and not in ServerContext
